[INDY-1792][INDY-1784][pool automation] Script to create an inventory directory with group_vars

andkononykhin opened the pull request:
The script:
- creates inventory directory
- explores available roles for defaults
- provides API to pass custom values for variables
- dumps config files per role as group_vars

Inventory's group_vars might be passed to molecule.

Additionally the PR fixes AWS resources cleanup routine:

- waiting until instance termination complete before removal of the linked security group

andkononykhin commented:
test this please

skhoroshavin commented on pool_automation/roles/aws_manage/defaults/main.yml:
```diff
@@ -30,3 +24,12 @@ aws_regions:
+
+instance_count: 4
+
+pa_aws_prefix: PA
```
AFAIK this is going to add another tag to EC2 instances, so probably this better be named as `aws_tag_prefix` to be consistent with `aws_tag_namespace` and `aws_tag_role`. Also `prefix` looks very general here, probably something like `owner` or `dept` or `purpose` would be better.

andkononykhin replied:
Yes, it is going but not in that PR and it would be likely something like `Project`. It wouldn't be a tag prefix but rather a separate tag itself. Here it is used only for resources where no tags are created (for now) and it's really just a prefix for names.

skhoroshavin commented on pool_automation/scripts/README.md:
````diff
@@ -0,0 +1,55 @@
+```
+
+So it is possible to place your inventory file(s) here (e.g. `inventory-dir/hosts`)
+and pass either the whole directory or an inventory file to [Ansible command line tools][2aceed7f].
````
In fact now passing the whole dir as inventory still has quirks because after running `pool_create` keys and ssh_config are also treated as host list, giving lots of warnings.

andkononykhin replied:
Fixed

skhoroshavin commented on pool_automation/scripts/inventory-init.py:
```diff
@@ -0,0 +1,184 @@
+            if args.get(_arg_name):
+                config[_p] = args[_arg_name]
+
+        _dump_config("{}/{}_config.yml".format(group_vars_dir, role), config, params['defaults'])
```
Having single config file per group instead of per role might be more manageable, at least now

andkononykhin replied:
But it would be less readable since defaults will be accumulated for all roles.

andkononykhin replied:
Done

skhoroshavin commented on pool_automation/pool_destroy.yml:
```diff
@@ -3,4 +3,6 @@
-    - { role: aws_manage, tag_namespace: test, tag_role: node, instance_count: 0 }
+    - role: aws_manage
+      vars:
+        instance_count: 0
```
This can be simplified to
```
  - role: aws_manage
    instance_count: 0
```

andkononykhin replied:
Thanks. Fixed

andkononykhin commented on pool_automation/roles/aws_manage/defaults/main.yml:
```diff
@@ -1,17 +1,11 @@
-aws_type: t2.micro
-aws_image: ami-027583e616ca104df
+aws_ec2_type: t2.micro
+aws_ami_id: ami-027583e616ca104df
```
Could you explain why?

andkononykhin replied:
`aws_ami_id` has been removed

andkononykhin commented on pool_automation/roles/aws_manage/library/stateful_set.py:
```diff
@@ -96,15 +98,22 @@ def get_tag(inst, name):
+
     for region in AWS_REGIONS:
+        _terminated = []
+        _launched = []
```
just to separate them from function's top level ones

andkononykhin replied:
Fixed

andkononykhin commented on pool_automation/roles/aws_manage/library/stateful_set.py:
```diff
@@ -96,15 +98,22 @@ def get_tag(inst, name):
 
 
 HostInfo = namedtuple('HostInfo', 'tag_id public_ip user')
+AWSWaiterInfo = namedtuple('AWSWaiterInfo', 'waiter kwargs')
```
I think, yes, we do. [EC2.Waiter API|https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Waiter.InstanceTerminated.wait] accepts a set of useful arguments (like Filter) that might be utilized in future

andkononykhin commented:
@skhoroshavin 

> This was a lot of work, however this still has two serious problems:
>pool_install.yml doesn't work with inventory created with inventory_init.py mypool and updated with >ansible-playbook -i mypool pool_create.yml
There were two reasons:
1. ssh configs polluted inventory directory (as you mentioned). Fixed
2. `pool_install` play used `nodes` group but it didn't have any relations with `pool_create` play vars. Fixed
>it's great that inventory_init.py looks for defaults in roles in very generic way, however it would be much >more convenient if it automatically updated aws_tag_namespace accordingly. Yes, it can be updated >manually or with yet another parameter to inventory_init, but it looks odd and I fear this workflow will be >error-prone

Fixed: namespace now is a basename of the inventory if the latter is defined (default is the same)

andkononykhin commented:
Test this please
